proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,237 @@
+internal format of `EpochRewardProof` is an array of entries,
+```
+[(epoch, (reward_amount, root_hash))]
+```
+, where each entry corresponds to the reward for
+one particular epoch. The maximum depth of the history are the same as the
+depth of stake history, which is 512. The `root_hash` of the merkle tree for
+all rewards are optional. It is for potential future works to distribute
+rewards as transactions.
+
+When reaching the end of reward computation phase, `EpochRewardProof` are
+updated with the reward computation result. A new entry will be pushed into the
+array. If the length of the history array reaches the maximum depth, the oldest
+entry will popped.
+
+```
+struct RewardHistoryEntry {
+   total: u64,                  // total rewards for this epoch
+   distributed: u64,            // already distributed reward amount
+   root_hash: Option<Hash>,     // root-hash of reward merkle tree
+}
+
+type EpochReward = (Epoch, RewardHistoryEntry);
+
+struct RewardHistory {
+   rewards: [EpochReward; 512],
+}
+```
+
+At the end of reward computation phase, a set of `EpochRewardsReserve` sysvar
+accounts will be updated too. Each `RewardsReserve` records the reward balance
+to be distributed for a particular block in `Reward Distribution` phase, which
+captures a particular subset of stakes rewards to be paid at the `index` block in
+the reward distribution phase. The complete subset of stake rewards forms a
+partition of the all the stake rewards.
+
+```
+struct EpochRewardReserve {
+   balance: u64,
+   index: u64,
+   hash: Hash
+}
+```
+
+`balance` is total amount of rewards (in lamports) to be distributed in this block.
+`index` is the reward distribution block height offset from reward computation end block.
+`hash` is the hash of all distributed rewards for this block.
+
+
+### Reward Distribution
+
+Reward distribution phase happens right after reward computation phase and
+lasts for `M` blocks. For each block, `i` in reward distribution phase, the
+rewards correspond to `EpochRewardReserve` sysvar account with `index=i` will
+be paid out, approximate `1/M` of the total stake reward accounts.
+
+This partition is required to be deterministic for the current epoch, but also
+random across different epochs. This can be achieved by hashing the account's
+pubkey with some epoch dependent values, sort the results, and divide them into
+`M` bins. The epoch dependent value can be the epoch number, total rewards for
+the epoch, the leader pubkey for the epoch block, etc. `M` are chosen based on
+50K account per block, which equal to `ceil(num_stake_accounts/50,000)`.
+
+
+### Restrict Stake Account access
+
+To avoid the complexity and potential rewards in zombie stake accounts, all
+write access to stake accounts are restricted. That means any update to stake
+accounts has to wait till the rewards distribution end. Any transactions
+involving stake programs during the epoch rewarding phase will fail with a unique transaction error code.
```

Comment:
> If zombie accounts are the issue, can we allow write access for instructions that don't close accounts?

The concern was around implementation complexity. It's _much_ cleaner and safer to just disable all stake program access until the rewards are distributed rather than considering all possible execution paths through the stake program and missing one.

It's expected that the stake program would only be inaccessible for a minute or so at worst. And the argument was made that this is effectively already the case at epoch boundaries since the chain stalls while rewards are paid. So the stake program is _already_ inaccessible and this proposal simply permits other transactions to execute during the time that the stake program is currently offline